ros2doctor/ros2doctor/api/package.py

```diff
@@ -0,0 +1,144 @@
+    repos_info = distro_data.get('repositories')
+    distro_package_vers = {}
+    if repos_info:
+        for _, info in repos_info.items():
+            release = info.get('release')
+            if release:
+                version = release.get('version')
+                packages = release.get('packages')
+                if packages:
+                    for p in packages:
+                        distro_package_vers[p] = version
+    return distro_package_vers
+
+
+def compare_versions(local_prefixes: dict,
+                     distro_packages: dict,
+                     isCheck=True) -> Tuple[List, dict]:
+    warning_msgs = []
+    report_dict = {}
+    missing_req = ''
+    missing_local = ''
+    for name, prefix in local_prefixes.items():
+        try:
+            required_ver_str = distro_packages.get(name)
+            required_ver = version.parse(required_ver_str).base_version
+        except TypeError:
+            missing_req += ' ' + name
+            required_ver = version.parse('').base_version
+        file_path = os.path.join(prefix, 'share', name)
+        try:
+            package_obj = parse_package(file_path)
+            local_ver_str = package_obj.version
+            local_ver = version.parse(local_ver_str).base_version
+        except (AttributeError, IOError, InvalidPackage):
+            missing_local += ' ' + name
+            local_ver = version.parse('').base_version
+        if isCheck:
+            if local_ver < required_ver:
+                warning_msgs.append(f'{name} is deprecated.'
+                                    f' local: {local_ver} <'
+                                    f' required: {required_ver}')
+        else:
+            report_dict[name] = f'required={required_ver}, local={local_ver}'
+    if missing_req:
+        warning_msgs.append('Cannot find required versions of packages:' + missing_req)
+    if missing_local:
+        warning_msgs.append('Cannot find local versions of packages:' + missing_local)
+    return warning_msgs, report_dict
+
+
+class PackageCheck(DoctorCheck):
+    """Check local package versions against release versions on rosdistro."""
+
+    def category(self):
+        return 'package'
+
+    def check(self):
+        """Check packages within the directory where command is called."""
+        result = Result()
+        local_prefixes = get_packages_with_prefixes()
+        if not local_prefixes:
+            result.add_error('ERROR: local package info is not found.')
+        try:
+            distro_package_vers = get_distro_package_versions()
+            if not distro_package_vers:
+                result.add_error('ERROR: distro packages info is not found.')
+        except (AttributeError, RuntimeError, URLError):
+            result.add_error('ERROR: Unable to fetch package information from rosdistro.')
+        if result.error != 0:
+            return result
+
+        warning_msgs, _ = compare_versions(local_prefixes,
```

Comment:
Rather than passing the package names and prefixes, I think it would better to write a separate function to produce a dict from local package names to versions. Then `compare_versions` can take the local package versions dict instead and I think the logic becomes simpler. E.g. (without error handling)

```python
local_package_versions = get_local_package_versions()
distro_package_versions = get_distro_package_version()
warning_msgs = compare_versions(local_package_versions, distro_package_versions)
```


`PackageReport` could instead call the function for getting package versions instead of `compare_versions`, and we would no longer need the "is check" parameter:

```python
local_package_versions = get_local_package_versions()
distro_package_versions = get_distro_package_versions()
# Produce report
for name, local_version in local_package_versions:
    report.add_to_report(name, f'required={distro_package_versions.get(name)}, local={local_version}')
```
^ untested code.